stdlib/REPL/src/REPLCompletions.jl

```diff
@@ -213,8 +214,27 @@ function complete_symbol(sym::String, @nospecialize(ffunc), context_module::Modu
                 end
             end
         end
+
+        # Also, try abstract-interpreting propertynames
+        thunk = context_module.eval(:(() -> propertynames($(ex))))
+        inf_params = Core.Compiler.InferenceParams(; assume_bindings_static=true)
+        frame = Core.Compiler.typeinf_frame(
+            Core.Compiler.NativeInterpreter(;inf_params),
+            first(methods(thunk)),
+            Tuple{typeof(thunk)},
+            Core.svec(),
+            true,
+        )
```

Comment:
There are some ways available to perform inference on a toplevel expression directly rather than wrapping it in a temporary function and performing inference on it.
This piece of code from JET.jl might be useful:
https://github.com/aviatesk/JET.jl/blob/9f8add4ce3d8ba528249ec02023711759abb00d3/src/toplevel/virtualprocess.jl#L1395-L1415
(ignore the `abstract_global_symbols!` stuff)